src/manager/checkpoint.rs

```diff
@@ -0,0 +1,160 @@
+// Copyright 2022-2023 Protocol Labs
+// SPDX-License-Identifier: MIT
+use std::collections::hash_map::RandomState;
+use std::collections::{HashMap, HashSet};
+use std::time::Duration;
+
+use anyhow::Result;
+use cid::Cid;
+use fil_actors_runtime::cbor;
+use futures_util::future::join_all;
+use fvm_shared::address::Address;
+use fvm_shared::clock::ChainEpoch;
+use fvm_shared::MethodNum;
+use ipc_gateway::Checkpoint;
+use ipc_sdk::subnet_id::SubnetID;
+use primitives::TCid;
+use tokio::time::sleep;
+
+use crate::config::Subnet;
+use crate::lotus::client::LotusJsonRPCClient;
+use crate::lotus::message::mpool::MpoolPushMessage;
+use crate::lotus::LotusClient;
+
+const CHAIN_HEAD_REQUEST_PERIOD: Duration = Duration::from_secs(10);
+
+/// Starts the checkpoint manager.
+#[allow(dead_code)]
+pub async fn start(subnets: HashMap<String, Subnet>) {
+    let managed_subnets = subnets_to_manage(subnets);
+
+    let mut futures = Vec::new();
+    for (child, parent) in managed_subnets {
+        futures.push(manage_subnet(child, parent));
+    }
+    join_all(futures).await;
```

Comment:
Not sure how relevant this comment is, but if we use `join_all` then this `start` is cannot be cancelled nor changed and it will run forever. Not sure if this will ever be subject to change.

Reply:
Once we have the config reloading in place, we will want to "unwatch" subnets and start watching new nodes. This whole code will need to be restructured to accommodate that.

Reply:
I updated this logic. The driving of the futures was pushed to a dedicated task and I've added a select to drive the check of shutdown signal and config changes.